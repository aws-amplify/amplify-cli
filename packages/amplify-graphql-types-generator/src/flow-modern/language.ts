import {
  GraphQLEnumType,
  GraphQLInputObjectType,
} from 'graphql';

import {
  CompilerOptions
} from '../compiler';

import { createTypeAnnotationFromGraphQLTypeFunction } from './helpers';

import * as t from '@babel/types';

export type ObjectProperty = {
  name: string,
  description?: string | null | undefined,
  annotation: t.FlowTypeAnnotation
}

export interface FlowCompilerOptions extends CompilerOptions {
  useFlowExactObjects: boolean
}

export default class FlowGenerator {
  options: FlowCompilerOptions
  typeAnnotationFromGraphQLType: Function

  constructor(compilerOptions: FlowCompilerOptions) {
    this.options = compilerOptions;

    this.typeAnnotationFromGraphQLType = createTypeAnnotationFromGraphQLTypeFunction(compilerOptions);
  }

  public enumerationDeclaration(type: GraphQLEnumType) {
    const { name, description } = type;
    const unionValues = type.getValues().map(({ value }) => {
      const type = t.stringLiteralTypeAnnotation();
      type.value = value;

      return type;
    });

    const typeAlias = t.exportNamedDeclaration(
      t.typeAlias(
        t.identifier(name),
        undefined,
        t.unionTypeAnnotation(unionValues)
      ),
      []
    );

    typeAlias.leadingComments = [{
      type: 'CommentLine',
      value: ` ${description}`
    } as t.CommentLine];

    return typeAlias;
  }

  public inputObjectDeclaration(inputObjectType: GraphQLInputObjectType) {
    const { name, description } = inputObjectType;

    const fieldMap = inputObjectType.getFields();
    const fields: ObjectProperty[] = Object.keys(inputObjectType.getFields())
      .map((fieldName: string) => {
        const field = fieldMap[fieldName];
        return {
          name: fieldName,
          annotation: this.typeAnnotationFromGraphQLType(field.type)
        }
      });

    const typeAlias = this.typeAliasObject(name, fields);

    typeAlias.leadingComments = [{
      type: 'CommentLine',
      value: ` ${description}`
    } as t.CommentLine]

    return typeAlias;
  }

  public objectTypeAnnotation(fields: ObjectProperty[], isInputObject: boolean = false) {
    const objectTypeAnnotation = t.objectTypeAnnotation(
      fields.map(({name, description, annotation}) => {
        if (annotation.type === "NullableTypeAnnotation") {
          t.identifier(name + '?')
        }

        const objectTypeProperty = t.objectTypeProperty(
          t.identifier(
            // Nullable fields on input objects do not have to be defined
            // as well, so allow these fields to be "undefined"
            (isInputObject && annotation.type === "NullableTypeAnnotation")
              ? name + '?'
              : name
          ),
          annotation
        );

        if (description) {
          objectTypeProperty.trailingComments = [{
            type: 'CommentLine',
            value: ` ${description}`
          } as t.CommentLine]
        }

        return objectTypeProperty;
      })
    );

    if (this.options.useFlowExactObjects) {
      objectTypeAnnotation.exact = true;
    }

    return objectTypeAnnotation;
  }

  public typeAliasObject(name: string, fields: ObjectProperty[]) {
    return t.typeAlias(
      t.identifier(name),
      undefined,
      this.objectTypeAnnotation(fields)
    );
  }

  public typeAliasObjectUnion(name: string, members: ObjectProperty[][]) {
    return t.typeAlias(
      t.identifier(name),
      undefined,
      t.unionTypeAnnotation(
        members.map(member => {
          return this.objectTypeAnnotation(member)
        })
      )
    )
  }

  public typeAliasGenericUnion(name: string, members: t.FlowTypeAnnotation[]) {
    return t.typeAlias(
      t.identifier(name),
      undefined,
      t.unionTypeAnnotation(members)
    );
  }

  public exportDeclaration(declaration: t.Declaration) {
    return t.exportNamedDeclaration(declaration, []);
  }

  public annotationFromScopeStack(scope: string[]) {
    return t.genericTypeAnnotation(
      t.identifier(
        scope.join('_')
      )
    );
  }
}
