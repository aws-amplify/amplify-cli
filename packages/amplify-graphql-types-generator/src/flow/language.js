import {
  join,
  wrap,
} from '../utilities/printing';

import { propertyDeclarations } from './codeGeneration';
import { typeNameFromGraphQLType } from './types';

import { pascalCase } from 'change-case';

export function typeDeclaration(generator, { interfaceName, noBrackets }, closure) {
  generator.printNewlineIfNeeded();
  generator.printNewline();
  generator.print(`export type ${ interfaceName } = `);
  generator.pushScope({ typeName: interfaceName });
  if (noBrackets) {
    generator.withinBlock(closure, '', '');
  } else {
    generator.withinBlock(closure, '{|', '|}');
  }
  generator.popScope();
  generator.print(';');
}

export function propertyDeclaration(generator, {
  fieldName,
  type,
  propertyName,
  typeName,
  description,
  isArray,
  isNullable,
  isArrayElementNullable,
  fragmentSpreads,
  isInput
}, closure, open = ' {|', close = '|}') {
  const name = fieldName || propertyName;

  if (description) {
    description.split('\n')
      .forEach(line => {
        generator.printOnNewline(`// ${line.trim()}`);
      })
  }

  if (closure) {
    generator.printOnNewline(name)
    if (isInput && isNullable) {
      generator.print('?')
    }
    generator.print(':')
    if (isNullable) {
      generator.print(' ?');
    }
    if (isArray) {
      if (!isNullable) {
        generator.print(' ');
      }
      generator.print(' Array<');
      if (isArrayElementNullable) {
        generator.print('?');
      }
    }

    generator.pushScope({ typeName: name });

    generator.withinBlock(closure, open, close);

    generator.popScope();

    if (isArray) {
      generator.print(' >');
    }

  } else {
    generator.printOnNewline(name)
    if (isInput && isNullable) {
      generator.print('?')
    }
    generator.print(`: ${typeName || typeNameFromGraphQLType(generator.context, type)}`);
  }
  generator.print(',');
}

export function propertySetsDeclaration(generator, property, propertySets, standalone = false) {
  const {
    description, fieldName, propertyName, typeName,
    isNullable, isArray, isArrayElementNullable
  } = property;
  const name = fieldName || propertyName;

  if (description) {
    description.split('\n')
      .forEach(line => {
        generator.printOnNewline(`// ${line.trim()}`);
      })
  }
  if (!standalone) {
    generator.printOnNewline(`${name}:`);
  }

  if (isNullable) {
    generator.print(' ?');
  }

  if (isArray) {
    generator.print('Array< ');
    if (isArrayElementNullable) {
      generator.print('?');
    }
  }

  generator.pushScope({ typeName: name });

  generator.withinBlock(() => {
    propertySets.forEach((propertySet, index, propertySets) => {
      generator.withinBlock(() => {
        propertyDeclarations(generator, propertySet);
      });
      if (index !== propertySets.length - 1) {
        generator.print(' |');
      }
    })
  }, '(', ')');

  generator.popScope();

  if (isArray) {
    generator.print(' >');
  }

  if (!standalone) {
    generator.print(',');
  }
}
