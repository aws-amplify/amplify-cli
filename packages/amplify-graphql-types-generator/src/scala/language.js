import {
  join,
  wrap,
} from '../utilities/printing';

export function comment(generator, comment) {
  const split = comment ? comment.split('\n') : [];
  if (split.length > 0) {
    generator.printOnNewline('/**')
    split.forEach(line => {
      generator.printOnNewline(` * ${line.trim()}`);
    });

    generator.printOnNewline(' */');
  }
}

export function packageDeclaration(generator, pkg) {
  generator.printNewlineIfNeeded();
  generator.printOnNewline(`package ${pkg}`);
  generator.popScope();
}

export function objectDeclaration(generator, { objectName, superclass, properties }, closure) {
  generator.printNewlineIfNeeded();
  generator.printOnNewline(`object ${objectName}` + (superclass ? ` extends ${superclass}` : ''));
  generator.pushScope({ typeName: objectName });
  generator.withinBlock(closure);
  generator.popScope();
}

export function caseClassDeclaration(generator, { caseClassName, description, superclass, params }, closure) {
  generator.printNewlineIfNeeded();
  comment(generator, description);
  generator.printOnNewline(`case class ${caseClassName}(${(params || []).map(v => v.name + ": " + v.type).join(', ')})` + (superclass ? ` extends ${superclass}` : ''));
  generator.pushScope({ typeName: caseClassName });
  generator.withinBlock(closure);
  generator.popScope();
}

export function propertyDeclaration(generator, { propertyName, typeName, description}, closure) {
  comment(generator, description);
  generator.printOnNewline(`val ${propertyName}: ${typeName} =`);
  generator.withinBlock(closure);
}

export function propertyDeclarations(generator, declarations) {
  declarations.forEach(o => {
    propertyDeclaration(generator, o);
  });
}

const reservedKeywords = new Set(
  'case', 'catch', 'class', 'def', 'do', 'else',
  'extends', 'false', 'final', 'for', 'if', 'match',
  'new', 'null', 'throw', 'trait', 'true', 'try', 'until',
  'val', 'var', 'while', 'with'
);

export function escapeIdentifierIfNeeded(identifier) {
  if (reservedKeywords.has(identifier)) {
    return '`' + identifier + '`';
  } else {
    return identifier;
  }
}
