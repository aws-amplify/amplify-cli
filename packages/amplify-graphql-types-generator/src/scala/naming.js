import { camelCase, pascalCase } from 'change-case';
import * as Inflector from 'inflected';

import {
  join
} from '../utilities/printing';

import {
  escapeIdentifierIfNeeded
} from './language';

import {
  typeNameFromGraphQLType
} from './types';

import {
  GraphQLError,
  GraphQLList,
  GraphQLNonNull,
  getNamedType,
  isCompositeType,
} from 'graphql';

export function enumCaseName(name) {
  return camelCase(name);
}

export function operationClassName(name) {
  return pascalCase(name);
}

export function caseClassNameForPropertyName(propertyName) {
  return pascalCase(Inflector.singularize(propertyName));
}

export function caseClassNameForFragmentName(fragmentName) {
  return pascalCase(fragmentName);
}

export function caseClassNameForInlineFragment(inlineFragment) {
  return 'As' + pascalCase(String(inlineFragment.typeCondition));
}

export function propertyFromField(context, field, namespace) {
  const name = field.name || field.responseName;
  const unescapedPropertyName = isMetaFieldName(name) ? name : camelCase(name)
  const propertyName = escapeIdentifierIfNeeded(unescapedPropertyName);

  const type = field.type;
  const isList = type instanceof GraphQLList || type.ofType instanceof GraphQLList
  const isOptional = field.isConditional || !(type instanceof GraphQLNonNull);
  const bareType = getNamedType(type);

  if (isCompositeType(bareType)) {
    const bareTypeName = join([
      namespace,
      escapeIdentifierIfNeeded(pascalCase(Inflector.singularize(name)))
    ], '.');
    const typeName = typeNameFromGraphQLType(context, type, bareTypeName, isOptional);
    return { ...field, propertyName, typeName, bareTypeName, isOptional, isList, isComposite: true };
  } else {
    const typeName = typeNameFromGraphQLType(context, type, undefined, isOptional);
    return { ...field, propertyName, typeName, isOptional, isList, isComposite: false };
  }
}

export function propertyFromInlineFragment(context, inlineFragment) {
  const structName = caseClassNameForInlineFragment(inlineFragment);
  const propertyName = camelCase(structName);
  const typeName = structName + '?'
  return { propertyName, typeName, structName, isComposite: true, ...inlineFragment };
}

export function propertyFromFragmentSpread(context, fragmentSpread) {
  const fragmentName = fragmentSpread;
  const fragment = context.fragments[fragmentName];
  if (!fragment) {
    throw new GraphQLError(`Cannot find fragment "${fragmentName}"`);
  }
  const propertyName = camelCase(fragmentName);
  const typeName = caseClassNameForFragmentName(fragmentName);
  return { propertyName, typeName, fragment, isComposite: true };
}

function isMetaFieldName(name) {
  return name.startsWith("__");
}