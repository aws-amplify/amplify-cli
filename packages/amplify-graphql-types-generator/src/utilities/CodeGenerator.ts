export interface BasicGeneratedFile {
  output: string
}


export class GeneratedFile<Scope = any> implements BasicGeneratedFile {
  scopeStack: Scope[] = [];
  indentWidth = 2;
  indentLevel = 0;
  startOfIndentLevel = false;

  public output = '';

  pushScope(scope: Scope) {
    this.scopeStack.push(scope);
  }

  popScope() {
    return this.scopeStack.pop();
  }

  get scope(): Scope {
    if (this.scopeStack.length < 1) throw new Error('No active scope');

    return this.scopeStack[this.scopeStack.length - 1];
  }

  print(string?: string) {
    if (string) {
      this.output += string;
    }
  }

  printNewline() {
    if (this.output) {
      this.print('\n');
      this.startOfIndentLevel = false;
    }
  }

  printNewlineIfNeeded() {
    if (!this.startOfIndentLevel) {
      this.printNewline();
    }
  }

  printOnNewline(string?: string) {
    if (string) {
      this.printNewline();
      this.printIndent();
      this.print(string);
    }
  }

  printIndent() {
    const indentation = ' '.repeat(this.indentLevel * this.indentWidth);
    this.output += indentation;
  }

  withIndent(closure: Function) {
    if (!closure) return;

    this.indentLevel++;
    this.startOfIndentLevel = true;
    closure();
    this.indentLevel--;
  }

  withinBlock(closure: Function, open = ' {', close = '}') {
    this.print(open);
    this.withIndent(closure);
    this.printOnNewline(close);
  }
}

export default class CodeGenerator<Context = any, Scope = any> {
  generatedFiles: { [fileName: string]: GeneratedFile<Scope> } = {};
  currentFile: GeneratedFile<Scope>;

  constructor(public context: Context) {
    this.currentFile = new GeneratedFile();
  }

  withinFile(fileName: string, closure: Function) {
    let file = this.generatedFiles[fileName];
    if (!file) {
      file = new GeneratedFile();
      this.generatedFiles[fileName] = file;
    }
    const oldCurrentFile = this.currentFile;
    this.currentFile = file;
    closure();
    this.currentFile = oldCurrentFile;
  }

  get output(): string {
    return this.currentFile.output;
  }

  pushScope(scope: Scope) {
    this.currentFile.pushScope(scope);
  }

  popScope() {
    this.currentFile.popScope();
  }

  get scope(): Scope {
    return this.currentFile.scope;
  }

  print(string?: string) {
    this.currentFile.print(string);
  }

  printNewline() {
    this.currentFile.printNewline();
  }

  printNewlineIfNeeded() {
    this.currentFile.printNewlineIfNeeded();
  }

  printOnNewline(string?: string) {
    this.currentFile.printOnNewline(string);
  }

  printIndent() {
    this.currentFile.printIndent();
  }

  withIndent(closure: Function) {
    this.currentFile.withIndent(closure);
  }

  withinBlock(closure: Function, open = ' {', close = '}') {
    this.currentFile.withinBlock(closure, open, close);
  }
}
