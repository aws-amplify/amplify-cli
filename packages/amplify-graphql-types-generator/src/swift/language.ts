import CodeGenerator from '../utilities/CodeGenerator';

import { join, wrap } from '../utilities/printing';

export interface Class {
  className: string;
  modifiers: string[];
  superClass?: string;
  adoptedProtocols?: string[];
}

export interface Struct {
  structName: string;
  adoptedProtocols?: string[];
  description?: string;
}

export interface Protocol {
  protocolName: string;
  adoptedProtocols?: string[];
}

export interface Property {
  propertyName: string;
  typeName: string;
  isOptional?: boolean;
  description?: string;
}

export function escapedString(string: string) {
  return string.replace(/"/g, '\\"').replace(/\n/g, '\\n');
}

// prettier-ignore
const reservedKeywords = new Set(['associatedtype', 'class', 'deinit', 'enum', 'extension',
  'fileprivate', 'func', 'import', 'init', 'inout', 'internal', 'let', 'open',
  'operator', 'private', 'protocol', 'public', 'static', 'struct', 'subscript',
  'typealias', 'var', 'break', 'case', 'continue', 'default', 'defer', 'do',
  'else', 'fallthrough', 'for', 'guard', 'if', 'in', 'repeat', 'return',
  'switch', 'where', 'while', 'as', 'Any', 'catch', 'false', 'is', 'nil',
  'rethrows', 'super', 'self', 'Self', 'throw', 'throws', 'true', 'try',
  'associativity', 'convenience', 'dynamic', 'didSet', 'final', 'get', 'infix',
  'indirect', 'lazy', 'left', 'mutating', 'none', 'nonmutating', 'optional',
  'override', 'postfix', 'precedence', 'prefix', 'Protocol', 'required', 'right',
  'set', 'Type', 'unowned', 'weak', 'willSet']);

export function escapeIdentifierIfNeeded(identifier: string) {
  if (reservedKeywords.has(identifier)) {
    return '`' + identifier + '`';
  } else {
    return identifier;
  }
}

export class SwiftGenerator<Context> extends CodeGenerator<Context, { typeName: string }> {
  constructor(context: Context) {
    super(context);
  }

  multilineString(string: string) {
    this.printOnNewline(`"${escapedString(string)}"`);
  }

  comment(comment?: string) {
    comment &&
      comment.split('\n').forEach(line => {
        this.printOnNewline(`/// ${line.trim()}`);
      });
  }

  deprecationAttributes(isDeprecated: boolean | undefined, deprecationReason: string | undefined) {
    if (isDeprecated !== undefined && isDeprecated) {
      deprecationReason = (deprecationReason !== undefined && deprecationReason.length > 0) ? deprecationReason : ""
      this.printOnNewline(`@available(*, deprecated, message: "${escapedString(deprecationReason)}")`)
    }
  }

  namespaceDeclaration(namespace: string | undefined, closure: Function) {
    if (namespace) {
      this.printNewlineIfNeeded();
      this.printOnNewline(`/// ${namespace} namespace`);
      this.printOnNewline(`public enum ${namespace}`);
      this.pushScope({ typeName: namespace });
      this.withinBlock(closure);
      this.popScope();
    } else {
      if (closure) {
        closure();
      }
    }
  }

  namespaceExtensionDeclaration(namespace: string | undefined, closure: Function) {
    if (namespace) {
      this.printNewlineIfNeeded();
      this.printOnNewline(`/// ${namespace} namespace`);
      this.printOnNewline(`public extension ${namespace}`);
      this.pushScope({ typeName: namespace });
      this.withinBlock(closure);
      this.popScope();
    } else {
      if (closure) {
        closure();
      }
    }
  }

  classDeclaration({ className, modifiers, superClass, adoptedProtocols = [] }: Class, closure: Function) {
    this.printNewlineIfNeeded();
    this.printOnNewline(wrap('', join(modifiers, ' '), ' ') + `class ${className}`);
    this.print(wrap(': ', join([superClass, ...adoptedProtocols], ', ')));
    this.pushScope({ typeName: className });
    this.withinBlock(closure);
    this.popScope();
  }

  structDeclaration({ structName, description, adoptedProtocols = [] }: Struct, closure: Function) {
    this.printNewlineIfNeeded();
    this.comment(description);
    this.printOnNewline(`public struct ${escapeIdentifierIfNeeded(structName)}`);
    this.print(wrap(': ', join(adoptedProtocols, ', ')));
    this.pushScope({ typeName: structName });
    this.withinBlock(closure);
    this.popScope();
  }

  propertyDeclaration({ propertyName, typeName, description }: Property) {
    this.comment(description);
    this.printOnNewline(`public var ${escapeIdentifierIfNeeded(propertyName)}: ${typeName}`);
  }

  propertyDeclarations(properties: Property[]) {
    if (!properties) return;
    properties.forEach(property => this.propertyDeclaration(property));
  }

  protocolDeclaration({ protocolName, adoptedProtocols }: Protocol, closure: Function) {
    this.printNewlineIfNeeded();
    this.printOnNewline(`public protocol ${protocolName}`);
    this.print(wrap(': ', join(adoptedProtocols, ', ')));
    this.pushScope({ typeName: protocolName });
    this.withinBlock(closure);
    this.popScope();
  }

  protocolPropertyDeclaration({ propertyName, typeName }: Property) {
    this.printOnNewline(`var ${propertyName}: ${typeName} { get }`);
  }

  protocolPropertyDeclarations(properties: Property[]) {
    if (!properties) return;
    properties.forEach(property => this.protocolPropertyDeclaration(property));
  }
}
