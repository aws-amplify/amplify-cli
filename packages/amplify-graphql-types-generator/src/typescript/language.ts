import { LegacyInlineFragment } from '../compiler/legacyIR';

import { propertyDeclarations } from './codeGeneration';
import { typeNameFromGraphQLType } from './types';

import CodeGenerator from '../utilities/CodeGenerator';
import { GraphQLType } from 'graphql';
import Maybe from 'graphql/tsutils/Maybe';

export interface Property {
  fieldName?: string;
  fieldType?: GraphQLType;
  propertyName?: string;
  type?: GraphQLType;
  description?: Maybe<string>;
  typeName?: string;
  isComposite?: boolean;
  isNullable?: boolean;
  fields?: any[];
  inlineFragments?: LegacyInlineFragment[];
  fragmentSpreads?: any;
  isInput?: boolean;
  isArray?: boolean;
  isArrayElementNullable?: boolean | null;
}

export function interfaceDeclaration(
  generator: CodeGenerator,
  { interfaceName, noBrackets }: { interfaceName: string; noBrackets?: boolean },
  closure: () => void
) {
  generator.printNewlineIfNeeded();
  generator.printNewline();
  generator.print(`export type ${interfaceName} = `);
  generator.pushScope({ typeName: interfaceName });
  if (noBrackets) {
    generator.withinBlock(closure, '', '');
  } else {
    generator.withinBlock(closure, '{', '}');
  }
  generator.popScope();
  generator.print(';');
}

export function propertyDeclaration(
  generator: CodeGenerator,
  { fieldName, type, propertyName, typeName, description, isInput, isArray, isNullable, isArrayElementNullable }: Property,
  closure?: () => void
) {
  const name = fieldName || propertyName;

  if (description) {
    description.split('\n').forEach(line => {
      generator.printOnNewline(`// ${line.trim()}`);
    });
  }

  if (closure) {
    generator.printOnNewline(name);

    if (isNullable && isInput) {
      generator.print('?');
    }
    generator.print(': ');

    if (isArray) {
      generator.print(' Array<');
    }
    generator.pushScope({ typeName: name });

    generator.withinBlock(closure);

    generator.popScope();

    if (isArray) {
      if (isArrayElementNullable) {
        generator.print(' | null');
      }
      generator.print(' >');
    }

    if (isNullable) {
      generator.print(' | null');
    }
  } else {
    generator.printOnNewline(name);
    if (isInput && isNullable) {
      generator.print('?');
    }
    generator.print(`: ${typeName || (type && typeNameFromGraphQLType(generator.context, type))}`);
  }
  generator.print(',');
}

export function propertySetsDeclaration(generator: CodeGenerator, property: Property, propertySets: Property[][], standalone = false) {
  const { description, fieldName, propertyName, isNullable, isArray, isArrayElementNullable } = property;
  const name = fieldName || propertyName;

  if (description) {
    description.split('\n').forEach(line => {
      generator.printOnNewline(`// ${line.trim()}`);
    });
  }

  if (!standalone) {
    generator.printOnNewline(`${name}: `);
  }

  if (isArray) {
    generator.print(' Array<');
  }

  generator.pushScope({ typeName: name });

  generator.withinBlock(
    () => {
      propertySets.forEach((propertySet, index, propertySets) => {
        generator.withinBlock(() => {
          propertyDeclarations(generator, propertySet);
        });
        if (index !== propertySets.length - 1) {
          generator.print(' |');
        }
      });
    },
    '(',
    ')'
  );

  generator.popScope();

  if (isArray) {
    if (isArrayElementNullable) {
      generator.print(' | null');
    }
    generator.print(' >');
  }

  if (isNullable) {
    generator.print(' | null');
  }

  if (!standalone) {
    generator.print(',');
  }
}

export function methodDeclaration(
  generator: CodeGenerator,
  {
    methodName,
    returnType,
    async,
    args,
  }: {
    methodName: string;
    returnType: string;
    async: boolean;
    args: Array<string>;
  },
  closure: () => void
) {
  generator.printNewline();
  if (async) generator.print('async ');
  generator.print(`${methodName}(${args.join(', ')}):${returnType}`);
  generator.pushScope({ methodName });
  generator.withinBlock(closure, '{', '}');
  generator.popScope();
}
