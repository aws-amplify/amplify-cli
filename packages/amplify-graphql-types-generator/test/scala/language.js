import { stripIndent } from 'common-tags';

import CodeGenerator from '../../src/utilities/CodeGenerator';

import { objectDeclaration, caseClassDeclaration, propertyDeclaration } from '../../src/scala/language';

describe('Scala code generation: Basic language constructs', function() {
  let generator;

  beforeEach(function() {
    generator = new CodeGenerator();
  });

  test(`should generate a object declaration`, function() {
    objectDeclaration(generator, { objectName: 'Hero' }, () => {
      propertyDeclaration(generator, { propertyName: 'name', typeName: 'String' }, () => {});
      propertyDeclaration(generator, { propertyName: 'age', typeName: 'Int' }, () => {});
    });

    expect(generator.output).toBe(stripIndent`
      object Hero {
        val name: String = {
        }
        val age: Int = {
        }
      }
    `);
  });

  test(`should generate a case class declaration`, function() {
    caseClassDeclaration(
      generator,
      { caseClassName: 'Hero', params: [{ name: 'name', type: 'String' }, { name: 'age', type: 'Int' }] },
      () => {}
    );

    expect(generator.output).toBe(stripIndent`
      case class Hero(name: String, age: Int) {
      }
    `);
  });

  test(`should generate nested case class declarations`, function() {
    caseClassDeclaration(
      generator,
      { caseClassName: 'Hero', params: [{ name: 'name', type: 'String' }, { name: 'age', type: 'Int' }] },
      () => {
        caseClassDeclaration(generator, { caseClassName: 'Friend', params: [{ name: 'name', type: 'String' }] }, () => {});
      }
    );

    expect(generator.output).toBe(stripIndent`
      case class Hero(name: String, age: Int) {
        case class Friend(name: String) {
        }
      }
    `);
  });

  test(`should handle multi-line descriptions`, () => {
    caseClassDeclaration(generator, { caseClassName: 'Hero', description: 'A hero' }, () => {
      propertyDeclaration(
        generator,
        { propertyName: 'name', typeName: 'String', description: `A multiline comment \n on the hero's name.` },
        () => {}
      );
      propertyDeclaration(
        generator,
        { propertyName: 'age', typeName: 'String', description: `A multiline comment \n on the hero's age.` },
        () => {}
      );
    });

    expect(generator.output).toMatchSnapshot();
  });
});
