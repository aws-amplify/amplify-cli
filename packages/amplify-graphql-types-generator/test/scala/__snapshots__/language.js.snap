// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scala code generation: Basic language constructs should handle multi-line descriptions 1`] = `
"/**
 * A hero
 */
case class Hero() {
  /**
   * A multiline comment
   * on the hero's name.
   */
  val name: String = {
  }
  /**
   * A multiline comment
   * on the hero's age.
   */
  val age: String = {
  }
}"
`;
